File: core/src/net/artux/pda/map/engine/systems/player/MissionsSystem.java
package net.artux.pda.map.engine.systems.player;

import com.badlogic.ashley.core.ComponentMapper;
import com.badlogic.ashley.core.Engine;
import com.badlogic.ashley.core.Entity;
import com.badlogic.ashley.core.Family;
import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Preferences;
import com.badlogic.gdx.assets.AssetManager;
import com.badlogic.gdx.audio.Sound;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Disposable;

import net.artux.pda.map.DataRepository;
import net.artux.pda.map.di.scope.PerGameMap;
import net.artux.pda.map.engine.components.PassivityComponent;
import net.artux.pda.map.engine.components.Position;
import net.artux.pda.map.engine.components.map.QuestComponent;
import net.artux.pda.map.engine.pathfinding.own.Connection;
import net.artux.pda.map.engine.pathfinding.own.Digraph;
import net.artux.pda.map.engine.pathfinding.own.DijkstraPathFinder;
import net.artux.pda.map.engine.pathfinding.own.GraphPath;
import net.artux.pda.map.engine.pathfinding.own.Node;
import net.artux.pda.map.engine.systems.BaseSystem;
import net.artux.pda.map.engine.systems.SoundsSystem;
import net.artux.pda.map.ui.blocks.MessagesPlane;
import net.artux.pda.map.utils.Mappers;
import net.artux.pda.model.map.GameMap;
import net.artux.pda.model.map.Point;
import net.artux.pda.model.quest.CheckpointModel;
import net.artux.pda.model.quest.MissionModel;
import net.artux.pda.model.quest.StoryModel;
import net.artux.pda.model.quest.story.ParameterModel;
import net.artux.pda.model.quest.story.StoryDataModel;

import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

import javax.inject.Inject;

@PerGameMap
public class MissionsSystem extends BaseSystem implements Disposable {

    private ComponentMapper<Position> pm = ComponentMapper.getFor(Position.class);
    private ComponentMapper<QuestComponent> qcm = ComponentMapper.getFor(QuestComponent.class);

    private final float pixelsPerMeter = 3f;
    private final DataRepository dataRepository;
    private final CameraSystem cameraSystem;
    private final SoundsSystem soundsSystem;
    private final MessagesPlane messagesPlane;
    private final Digraph<GameMap> mapDigraph;
    private final DijkstraPathFinder<GameMap> pathFinder;
    private final Sound missionUpdatedSound;

    private MissionModel activeMission;
    private Vector2 targetPosition;

    @Inject
    public MissionsSystem(MessagesPlane messagesPlane, AssetManager assetManager, DataRepository dataRepository, SoundsSystem soundsSystem, CameraSystem cameraSystem) {
        super(Family.all(Position.class, QuestComponent.class).exclude(PassivityComponent.class).get());
        this.dataRepository = dataRepository;
        this.cameraSystem = cameraSystem;
        this.messagesPlane = messagesPlane;
        this.soundsSystem = soundsSystem;

        missionUpdatedSound = assetManager.get("audio/sounds/pda/pda_objective.ogg");

        mapDigraph = buildSystemFromStory();
        pathFinder = new DijkstraPathFinder<>();
    }

    @Override
    public void addedToEngine(Engine engine) {
        super.addedToEngine(engine);
        setActiveMissionByName("");
        loadPreferences();
    }

    public void updateData(StoryDataModel oldDataModel) {
        Set<String> oldParams = getParams(oldDataModel);
        Set<String> params = getParams(dataRepository.getStoryDataModel());

        params.removeIf(oldParams::contains);
        String[] paramArr = params.toArray(new String[]{});
        List<MissionModel> updatedMissions = getMissions(paramArr);
        for (MissionModel m :
                updatedMissions) {
            CheckpointModel checkpointModel = m.getCurrentCheckpoint(paramArr);
            messagesPlane.addMessage("avatars/a0.jpg", "Задание обновлено: " + m.getTitle(),
                    "Новая цель: " + checkpointModel.getTitle(), MessagesPlane.Length.SHORT);
            soundsSystem.playSound(missionUpdatedSound);
        }
    }

    public Set<String> getParams(StoryDataModel dataModel) {
        return dataModel.getParameters().stream()
                .map(ParameterModel::getKey).collect(Collectors.toSet());
    }

    public String[] getParams() {
        StoryDataModel dataModel = dataRepository.getStoryDataModel();
        return dataModel.getParameters().stream()
                .map(ParameterModel::getKey).toArray(String[]::new);
    }

    public List<MissionModel> getMissions() {
        return getMissions(getParams());
    }

    public List<MissionModel> getMissions(String[] params) {
        return dataRepository.getStoryModel().getCurrentMissions(params);
    }

    public List<QuestComponent> getPoints() {
        LinkedList<QuestComponent> points = new LinkedList<>();
        for (Entity entity : getEntities()) {
            points.add(qcm.get(entity));
        }
        return points;
    }

    public void setActiveMissionByName(String name) {
        List<MissionModel> missionModels = getMissions();
        for (MissionModel m : missionModels) {
            if (m.getName().equals(name)) {
                setActiveMission(activeMission);
            }
        }
        if (activeMission == null) {
            if (missionModels.size() > 0)
                setActiveMission(missionModels.get(0));
            else if (getEntities().size() > 0)
                setTargetPosition(pm.get(getEntities().first()));
        }
    }

    public MissionModel getActiveMission() {
        return activeMission;
    }

    public void setActiveMission(MissionModel activeMission) {
        this.activeMission = activeMission;

        if (activeMission == null)
            return;

        CheckpointModel currentCheckpoint = activeMission.getCurrentCheckpoint(getParams());
        int chapter = currentCheckpoint.getChapter();
        int stage = currentCheckpoint.getStage();
        boolean found = false;
        for (Entity questEntity : getEntities()) {
            QuestComponent questComponent = qcm.get(questEntity);
            Position position = pm.get(questEntity);
            if (questComponent.contains(chapter, stage)) {
                found = true;
                setTargetPosition(position);
            }
        }
        if (!found) {
            GameMap currentMap = dataRepository.getGameMap();
            GameMap requiredMap = getRequiredMap(chapter, stage);
            if (requiredMap == currentMap || requiredMap == null)
                return;

            GraphPath<GameMap> path = pathFinder.find(mapDigraph, mapDigraph.offer(currentMap), mapDigraph.offer(requiredMap));
            if (path == null)
                return;

            Optional<Connection<GameMap>> connection = path.getConnections().stream().findFirst();
            connection.ifPresent(gameMapConnection -> setTargetPosition(Mappers
                    .vector2(((Point) gameMapConnection.getUserObject()).getPos())));
        }

    }

    @Override
    public void update(float deltaTime) {
        super.update(deltaTime);
    }

    @Override
    protected void processEntity(Entity entity, float deltaTime) {

    }

    private GameMap getRequiredMap(int chapter, int stage) {
        StoryModel storyModel = dataRepository.getStoryModel();
        for (GameMap map : storyModel.getMaps().values()) {
            for (Point point : map.getPoints()) {
                Map<String, String> currentData = point.getData();
                String chapterString = currentData.get("chapter");
                String stageString = currentData.get("stage");
                if (chapterString != null && stageString != null) {
                    int stageId = Integer.parseInt(stageString);
                    int chapterId = Integer.parseInt(chapterString);
                    if (chapterId == chapter && stageId == stage)
                        return map;
                }
            }
        }
        return null;
    }

    private Digraph<GameMap> buildSystemFromStory() {
        StoryModel storyModel = dataRepository.getStoryModel();
        Digraph<GameMap> mapDigraph = new Digraph<>();
        for (GameMap map : storyModel.getMaps().values()) {
            Node<GameMap> node = mapDigraph.offer(map);
            for (Point point : map.getPoints()) {
                if (point.getType() == 7) {
                    Map<String, String> currentData = point.getData();
                    String chapterString = currentData.get("chapter");
                    String stageString = currentData.get("stage");
                    String targetMap = currentData.get("map");//with map
                    if (chapterString != null && stageString != null) {
                        //with chapter
                        int stage = Integer.parseInt(stageString);
                        Map<String, String> data = storyModel
                                .getChapter(chapterString)
                                .getStage(stage).getData();
                        if (data != null && data.containsKey("map"))
                            targetMap = data.get("map");
                    }
                    if (targetMap != null) {
                        GameMap targetGameMap = storyModel.getMap(Long.parseLong(targetMap));
                        if (targetGameMap != null) {
                            Node<GameMap> gameMapNode = mapDigraph.offer(targetGameMap);
                            Connection<GameMap> connection = node.addConnection(gameMapNode, 1);
                            connection.setUserObject(point);
                        }
                    }
                }
            }
        }
        return mapDigraph;
    }

    public Vector2 getPosition() {
        Position playerPosition = pm.get(getPlayer());
        return playerPosition.getPosition();
    }

    public int getTargetDistance() {
        if (targetPosition != null) {
            if (!targetPosition.isZero()) {
                return (int) (getPosition().dst(targetPosition) / pixelsPerMeter);
            }
        }
        return -1;
    }

    public double getTargetAngle() {
        if (targetPosition != null) {
            if (!targetPosition.isZero()) {
                Vector2 pos = getPosition();
                return Math.atan2(
                        pos.y - targetPosition.y,
                        pos.x - targetPosition.x
                ) * 180.0d / Math.PI + 90;
            }
        }
        return 0;
    }

    public void setTargetPosition(Vector2 position) {
        if (position != null) {
            targetPosition = position;

            cameraSystem.setDetached(true);
            cameraSystem.getCamera().position.x = position.x;
            cameraSystem.getCamera().position.y = position.y;
        }
    }

    @Override
    public void dispose() {
        savePreferences();
    }

    public void savePreferences() {
        if (getPlayer() != null) {
            Preferences preferences = Gdx.app.getPreferences("missions");
            if (activeMission != null)
                preferences.putString("active", activeMission.getName());
            preferences.flush();
        }
    }

    private void loadPreferences() {
        if (getPlayer() != null) {
            Preferences preferences = Gdx.app.getPreferences("missions");

            String name = preferences.getString("active");
            setActiveMissionByName(name);
        }
    }

}

File: core/src/net/artux/pda/map/engine/ecs/systems/player/MissionsSystem.kt
package net.artux.pda.map.engine.ecs.systems.player

import com.badlogic.ashley.core.ComponentMapper
import com.badlogic.ashley.core.Engine
import com.badlogic.ashley.core.Entity
import com.badlogic.ashley.core.Family
import com.badlogic.gdx.Gdx
import com.badlogic.gdx.assets.AssetManager
import com.badlogic.gdx.audio.Sound
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.utils.Disposable
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import net.artux.engine.pathfinding.own.Connection
import net.artux.engine.pathfinding.own.Digraph
import net.artux.engine.pathfinding.own.DijkstraPathFinder
import net.artux.pda.map.DataRepository
import net.artux.pda.map.engine.ecs.components.BodyComponent
import net.artux.pda.map.engine.ecs.components.PassivityComponent
import net.artux.pda.map.engine.ecs.components.map.QuestComponent
import net.artux.pda.map.engine.ecs.systems.BaseSystem
import net.artux.pda.map.engine.ecs.systems.SoundsSystem
import net.artux.pda.map.utils.Mappers
import net.artux.pda.map.utils.di.scope.PerGameMap
import net.artux.pda.map.view.blocks.MessagesPlane
import net.artux.pda.model.map.GameMap
import net.artux.pda.model.map.Point
import net.artux.pda.model.quest.MissionModel
import net.artux.pda.model.quest.story.ParameterModel
import net.artux.pda.model.quest.story.StoryDataModel
import java.util.LinkedList
import java.util.stream.Collectors
import javax.inject.Inject

@PerGameMap
class MissionsSystem @Inject constructor(
    private val messagesPlane: MessagesPlane,
    assetManager: AssetManager,
    private val dataRepository: DataRepository,
    private val soundsSystem: SoundsSystem,
    private val cameraSystem: CameraSystem
) : BaseSystem(
    Family.all(
        BodyComponent::class.java, QuestComponent::class.java
    ).exclude(
        PassivityComponent::class.java
    ).get()
), Disposable {

    private val pm = ComponentMapper.getFor(
        BodyComponent::class.java
    )

    private val qcm = ComponentMapper.getFor(
        QuestComponent::class.java
    )

    private val pixelsPerMeter = 3f
    private val mapDigraph: Digraph<GameMap>
    private val pathFinder: DijkstraPathFinder<GameMap>
    private val missionUpdatedSound: Sound
    private var currentStoryDataModel: StoryDataModel
    private var activeMission: MissionModel? = null
    private var targetPosition: Vector2? = null

    init {
        missionUpdatedSound = assetManager.get("audio/sounds/pda/pda_objective.ogg")
        mapDigraph = buildSystemFromStory()
        pathFinder = DijkstraPathFinder()

        CoroutineScope(Dispatchers.Main).launch {
            dataRepository.storyDataModelFlow.collect {
                currentStoryDataModel = it
                updateData(dataRepository.storyDataModel)
            }
        }

        currentStoryDataModel = dataRepository.storyDataModel
    }

    override fun addedToEngine(engine: Engine) {
        super.addedToEngine(engine)
        setActiveMissionByName("")
        loadPreferences()
    }

    fun updateData(oldDataModel: StoryDataModel) {
        val paramSet = getUpdatedParams(oldDataModel)
        val paramArr = paramSet.toTypedArray()
        val updatedMissions = getMissions(paramArr)
        for (m in updatedMissions) {
            val checkpointModel = m.getCurrentCheckpoint(*paramArr)
            messagesPlane.addMessage(
                "textures/avatars/a0.png", "Задание обновлено: " + m.title,
                "Новая цель: " + checkpointModel!!.title, MessagesPlane.Length.SHORT
            )
            soundsSystem.playSound(missionUpdatedSound)
        }
    }

    fun getParams(dataModel: StoryDataModel): MutableSet<String> {
        return dataModel.parameters.stream()
            .map { obj: ParameterModel -> obj.key }.collect(Collectors.toSet())
    }

    fun getUpdatedParams(oldDataModel: StoryDataModel): Set<String> {
        val oldParams: Set<String> = getParams(oldDataModel)
        val params = getParams(currentStoryDataModel)
        params.removeIf { o: String -> oldParams.contains(o) }
        return params
    }

    val params: Array<String>
        get() {
            val dataModel = currentStoryDataModel
            return dataModel.parameters.stream()
                .map { obj: ParameterModel -> obj.key }.collect(Collectors.toSet()).toTypedArray()
        }
    val missions: List<MissionModel>
        get() = getMissions(params)

    fun getMissions(params: Array<String>): List<MissionModel> {
        return dataRepository.storyModel.getCurrentMissions(*params)
    }

    val points: List<QuestComponent>
        get() {
            val points = LinkedList<QuestComponent>()
            for (entity in entities) {
                points.add(qcm[entity])
            }
            return points
        }

    fun setActiveMissionByName(name: String) {
        val missionModels = missions
        for (m in missionModels) {
            if (m.name == name) {
                setActiveMission(activeMission)
            }
        }
        if (activeMission == null) {
            if (missionModels.isNotEmpty())
                setActiveMission(missionModels[0])
            else if (entities.size() > 0)
                setTargetPosition(pm[entities.first()].position)
        }
    }

    fun getActiveMission(): MissionModel? {
        return activeMission
    }

    fun setActiveMission(activeMission: MissionModel?) {
        this.activeMission = activeMission
        if (activeMission == null) return
        val currentCheckpoint = activeMission.getCurrentCheckpoint(*params)
        val chapter = currentCheckpoint?.chapter
        val stage = currentCheckpoint?.stage
        var found = false
        for (questEntity in entities) {
            val questComponent = qcm[questEntity]
            val position = pm[questEntity]
            if (questComponent.contains(chapter!!, stage!!)) {
                found = true
                setTargetPosition(position.position)
            }
        }
        if (!found) {
            val currentMap = dataRepository.gameMap
            val requiredMap = getRequiredMap(chapter!!, stage!!)
            if (requiredMap === currentMap || requiredMap == null) return
            val path = pathFinder.find(
                mapDigraph,
                mapDigraph.offer(currentMap),
                mapDigraph.offer(requiredMap)
            )
                ?: return
            val connection = path.connections.stream().findFirst()
            connection.ifPresent { gameMapConnection: Connection<GameMap> ->
                setTargetPosition(
                    Mappers
                        .vector2((gameMapConnection.userObject as Point).pos)
                )
            }
        }
    }

    override fun processEntity(entity: Entity, deltaTime: Float) {}

    private fun getRequiredMap(chapter: Int, stage: Int): GameMap? {
        val storyModel = dataRepository.storyModel
        for (map in storyModel.maps.values) {
            if (map.points != null)
                for (point in map.points!!) {
                    val currentData: Map<String, String> = point.data
                    val chapterString = currentData["chapter"]
                    val stageString = currentData["stage"]
                    if (chapterString != null && stageString != null) {
                        val stageId = stageString.toInt()
                        val chapterId = chapterString.toInt()
                        if (chapterId == chapter && stageId == stage) return map
                    }
                }
        }
        return null
    }

    private fun buildSystemFromStory(): Digraph<GameMap> {
        val storyModel = dataRepository.storyModel
        val mapDigraph = Digraph<GameMap>()
        for (map in storyModel.maps.values) {
            val node = mapDigraph.offer(map)
            for (point in map.points!!) {
                if (point.type == 7) {
                    val currentData: HashMap<String, String>? = point.data
                    val chapterString = currentData?.get("chapter")
                    val stageString = currentData?.get("stage")
                    var targetMap = currentData?.get("map") //with map
                    if (chapterString != null && stageString != null) {
                        //with chapter
                        val stage = stageString.toLong()
                        val data: Map<String, String>? = storyModel
                            .getChapter(chapterString)
                            ?.getStage(stage)?.data
                        if (data != null && data.containsKey("map")) targetMap = data["map"]
                    }
                    if (targetMap != null) {
                        val targetGameMap = storyModel.getMap(targetMap.toLong())
                        if (targetGameMap != null) {
                            val gameMapNode = mapDigraph.offer(targetGameMap)
                            val connection = node.addConnection(gameMapNode, 1)
                            connection.userObject = point
                        }
                    }
                }
            }
        }
        return mapDigraph
    }

    val position: Vector2
        get() {
            val playerPosition = pm[player]
            return playerPosition.position
        }
    val targetDistance: Int
        get() {
            if (targetPosition != null) {
                if (!targetPosition!!.isZero) {
                    return (position.dst(targetPosition) / pixelsPerMeter).toInt()
                }
            }
            return -1
        }
    val targetAngle: Double
        get() {
            if (targetPosition != null) {
                if (!targetPosition!!.isZero) {
                    val pos = position
                    return Math.atan2(
                        (
                                pos.y - targetPosition!!.y).toDouble(),
                        (
                                pos.x - targetPosition!!.x
                                ).toDouble()
                    ) * 180.0 / Math.PI + 90
                }
            }
            return 0.0
        }

    fun setTargetPosition(position: Vector2?) {
        if (position != null) {
            targetPosition = position
            cameraSystem.isDetached = true
            cameraSystem.camera.position.x = position.x
            cameraSystem.camera.position.y = position.y
        }
    }

    override fun dispose() {
        savePreferences()
    }

    fun savePreferences() {
        if (player != null) {
            val preferences = Gdx.app.getPreferences("missions")
            if (activeMission != null) preferences.putString("active", activeMission!!.name)
            preferences.flush()
        }
    }

    private fun loadPreferences() {
        if (player != null) {
            val preferences = Gdx.app.getPreferences("missions")
            val name = preferences.getString("active")
            setActiveMissionByName(name)
        }
    }
}